### ISSUE TITLE
Question: about the decompress speed on Arm

### ISSUE THREAD

**Issue by chirping78** [zip-rs/zip#243]
A very simple program very like the example:
```
                for i in 0..archive.len() {
                    let mut file = archive.by_index(i).unwrap();
                    let outpath = file.enclosed_name().unwrap();
                    let mut outfile = std::fs::File::create(parent.join(outpath)).unwrap();
                    debug!("extract: {} ==> {}", outpath.display(), parent.join(outpath).display());
                    std::io::copy(&mut file, &mut outfile).unwrap();
                }

```
I use this program to decompress a file with about 400MB size, which contains 2~3 200MB files.
The speed is normal on X86 pc, but is very very slow on an Arm board.
On this Arm board, the busybox unzip takes about 20s to decompress, but the rust version has run about 20 minitues, still not finished...

**Comment by flxo** [comment]
@chirping78 can you post the output of strace when running your test? e.g `strace -f -s 256 <your binary>.

I've seen a lot 1 byte reads done by zip-rs when reading a file from the archive - even when the reader is a `BufRead`.

**Comment by chirping78** [comment]
I can reproduce the issue with the original example/extract:
```
 cargo b --target=armv7-unknown-linux-gnueabihf --example extract
```
when using the `extract` to decompress the file, the output speed is about 200kB / sec.
Since I'm using armv7 as the target, I think you can also reproduce it on raspberrypi, beagle, etc...

**Comment by chirping78** [comment]
@flex Ok, will try your suggestion after short while...

> @chirping78 can you post the output of strace when running your test? e.g `strace -f -s 256 .
> 
> I've seen a lot 1 byte reads done by zip-rs when reading a file from the archive - even when the reader is a `BufRead`.

**Comment by chirping78** [comment]
I have captured the strace log, it's too long to paste here, but the pattern is repeated. Paste some part here:
```
read(3, "\375\277\373\357\227 ....)=32768
write(4, "\2271\357\316?\307\   ...)=8192
write(4, "\3713\365<\224\207B  ...)=8192
write(4, "]\252\343\10\326\17
write(4, "\23\320N\235l\322:\
write(4, "\345\252\363j\340\2
read(3, ";}{\255\274)d\264\37
write(4, "\2019\347\6\37\2469
write(4, "\303\210M_\266.\217
write(4, "gVhM\235\320\357\21
write(4, "\370n\372\331\244\3
write(4, "f{\356UB\316b\26\33
read(3, "Q|\300\274~\360PB'\2
write(4, "s\27\17\3701nK4\363
write(4, "L\v\360\265\35\344k
write(4, "'\27(\377\f\372tI\3
write(4, ".\332\255\304\27\37
write(4, "u\331\0.\26\255\363
read(3, "\307@\336\255\360\24
write(4, "\211\242\355\216\26
write(4, "\225\356 \340\351\3
write(4, "\336\373\177H\333'\
write(4, "\360\2616\302W[&r0\
write(4, "\332[\37.K~H8\207}o
read(3, "\307\241\7\244\367\3
write(4, "\4\7\340\302\270\20
write(4, "\261^^\27\257\27|%\
write(4, "\371\314\317\347\
```
`3` is the input zip file, `4` is the output file.

**Comment by chirping78** [comment]
Looks like the  input/output buffer does not match the other side ??

**Comment by flxo** [comment]
@chirping78 That looks different from what I've seen here:

**Comment by chirping78** [comment]
@flxo your problem is different from mine. Your program is seeking in the archive to read comments. In my situation, it's in the decompressing stage to read/write the file. Anyway, I will compare with X86 pc to see the difference.

**Comment by Plecra** [comment]
I'd expect this to be a problem with the actual decompression of the file. Can you try using `miniz_oxide` to compress and decompress some data in memory?

**Comment by pixiaopai** [comment]
> I'd expect this to be a problem with the actual decompression of the file. Can you try using `miniz_oxide` to compress and decompress some data in memory?

flate2 = { version = "1.0.0", default-features = false, optional = true }
i try to this feature 
flate2 = { version = "1.0.0", features = ["zlib"], default-features = false }
Compile using
 cargo b --target=armv7-unknown-linux-gnueabihf --example extract
when using the extract to decompress the file, the output speed is normal. 

How should I use this library, need to modify this feature？
I tried the following ways
 use lib for zip-rs, but the speed is slow, Only used in the project to compile extract

**Comment by zamazan4ik** [comment]
Honestly, not sure, what should we do with this issue now. Maybe some people with Arm hardware can help us with the reproducibility of the issue?

**Comment by Plecra** [comment]
Yup, that's about where I land. This would be great to fix, but we need someone to come along with some Arm hardware who can help.

The first thing to do is testing the speed of the `miniz_oxide` compression on arm, to see if this is an upstream issue. If not, we'll need to find the bottleneck so that we can improve its performance.

**Comment by Plecra** [comment]
This is a major issue, but we don't have systems that we can use to fix the issue. I'm going to leave this open until we find someone able to work on the issue or a way to test `zip` on the ARM platform.

If anyone is suffering from this issue, please give us a full rundown of the OS, CPU, and zip archive you're trying to extract :) I want to be able to repro the problem! (Oh, and if you know a way I could use a cloud environment to debug this, I'd love to hear about it_

**Comment by zamazan4ik** [comment]
Actually, I could help to test it since I have M1-based Macbook :) So if you can share the problematic archive and reproduction steps - I will test it locally on my ARM machine and report the results.

**Comment by Pr0methean** [comment]
@zamazan4ik @chirping78  Does https://crates.io/crates/zip_next perform better on your test boards?
